### ISSUE TITLE
Race condition in ReactiveCommand.CanExecuteObservable

### ISSUE THREAD

**Issue by kentcb** [reactiveui/ReactiveUI#727]
Hi,

Just stumbled upon a race condition demonstrated by this test:

```C#
[Test]
public async Task REPRO()
{
    using (var canExecute = new BehaviorSubject<bool>(false))
    using (var command = ReactiveCommand.CreateAsyncTask(canExecute, _ => Task.FromResult(true)))
    {
        // this always works (of course)
        //await Task.Run(() => canExecute.OnNext(true));

        // this seems to never work (lack of delay exacerbates the problem)
        //Task.Run(() => canExecute.OnNext(true));

        // this sometimes work, sometimes not (lower delays decrease chance of success)
        Task.Run(() =>
        {
            Thread.Sleep(10);
            canExecute.OnNext(true);
        });

        await command.CanExecuteObservable
            .Where(x => x)
            .Timeout(TimeSpan.FromSeconds(3))
            .FirstAsync();
    }
}
```

**Comment by kentcb** [comment]
OK, it's been a while, but I think I know what you're saying now. Basically we could be unnecessarily re-scheduling the message when it's already on the UI thread?

This makes me wonder whether a better approach might be to add diagnostics to the implementation such that any tick from the `canExecute` observable is guarded. This would require some means of accurately determining the UI thread, which is a problem I brought up on Slack recently. There are several places that could benefit from knowing whether they're on the UI thread or not, like property change notifications.

I wish Rx had some kind of `EnsureObserveOn` operator that only re-schedules if necessary (i.e. not already in the correct context). If we could reliably build our own version of that operator, I think we could massively widen the pit of success for users by trapping and/or automatically handling common problems.

**Comment by kentcb** [comment]
I think we got a bit side-tracked with my signalling from the wrong thread. Even when using the scheduler there was a race. However, I believe this to be fixed by #971, which will be merged shortly. Hence, closing this issue for now.
